libcob/profiling.c

```diff
@@ -0,0 +1,526 @@
+ * to be monotonic, by using an internal storage of the previous
+ * time. */
+static cob_ns_time
+get_ns_time (void)
+{
+	static cob_ns_time ns_time = 0;
+	if (is_test){
+		ns_time += 1000000;
+		return ns_time;
+	} else {
+		unsigned long long nanoseconds;
+#ifdef _WIN32
+		LARGE_INTEGER performance_counter;
+		LARGE_INTEGER performance_frequency;
+		QueryPerformanceCounter(&performance_counter);
+		QueryPerformanceFrequency(&performance_frequency);
+		performance_counter.QuadPart *= 1000000000;
+		performance_counter.QuadPart /= performance_frequency.QuadPart;
+		nanoseconds = performance_counter.QuadPart;
+		if (nanoseconds>ns_time) ns_time = nanoseconds;
+		return ns_time;
+#else
+#if defined (HAVE_CLOCK_GETTIME)
+		struct timespec ts;
+		clock_gettime(CLOCK_MONOTONIC, &ts);
+		nanoseconds = ts.tv_sec * 1000000000 + ts.tv_nsec;
+#else
+		nanoseconds = clock() * 1000000000 / CLOCKS_PER_SEC;
+#endif /* HAVE_CLOCK_GETTIME */
+#endif /* _WIN32 */
+		if (nanoseconds>ns_time) ns_time = nanoseconds;
+		return ns_time;
+	}
+}
+
+static void
+prof_init_static ()
+{
+	static int init_done = 0;
+
+	if (!init_done && cobsetptr){
+		init_done = 1;
+		is_active = cobsetptr->cob_prof_enable;
+		if (is_active){
+			is_test = !!getenv ("COB_IS_RUNNING_IN_TESTMODE");
+		}
+	}
+}
+
+void cob_init_prof (cob_global *lptr, cob_settings *sptr)
+{
+	cobglobptr = lptr;
+        cobsetptr  = sptr;
+}
+
+struct cob_prof_module *
+cob_prof_init_module (cob_module *module,
+		      struct cob_prof_procedure *procedures,
+		      size_t procedure_count)
+{
+	prof_init_static();
+	if (is_active){
+		struct cob_prof_module *info;
+		struct cob_prof_module_list *item;
+
+		info = cob_malloc (sizeof(struct cob_prof_module));
+		info->total_times = cob_malloc ( procedure_count * sizeof(cob_ns_time) );
+		info->called_count = cob_malloc ( procedure_count * sizeof(unsigned int) );
+		info->procedure_recursions = cob_malloc ( procedure_count * sizeof(unsigned int) );
+		info->procedures = procedures;
+		info->procedure_count = procedure_count;
+
+		item = cob_malloc (sizeof(struct cob_prof_module_list));
+		item->info = info;
+		item->next = prof_info_list;
+		prof_info_list = item;
+		return info;
+	}
+	return NULL;
+}
+
+static void
+cob_prof_realloc_arrays (void)
+{
+	int new_size = max_prof_depth * 2 + 16;
+
+	if (new_size > cobsetptr->cob_prof_max_depth)
+		new_size = cobsetptr->cob_prof_max_depth;
+
+	if (max_prof_depth >= new_size){
+		int i;
+		fprintf (stderr, _("[cob_prof] Profiling overflow at %d calls, aborting profiling.\n"), current_idx);
+		fprintf (stderr, _("  Last 10 calls on stack:\n"));
+		for (i=0;i<10;i++){
+			struct cob_prof_module *info = called_runtimes[current_idx-1-i];
+			int proc_idx = called_procedures[current_idx-1-i];
+			struct cob_prof_procedure *proc = info->procedures + proc_idx;
+			fprintf (stderr, _("  * %s at %s:%d\n"), proc->text,
+				 proc->file, proc->line);
+		}
+		is_active = 0;
+		return;
+	}
```

Comment:
we should add a testcase of this and the output should use `cob_runtime_warning (` instead of `fprintf (stderr,`